File: VB/CustomItemExtension/CustomItems/WebPageItem/WebPageItemControlProvider.vb
Imports DevExpress.DashboardCommon
Imports DevExpress.DashboardWin
Imports DevExpress.XtraPrinting.Drawing
Imports DevExpress.XtraReports.UI
Imports Microsoft.Web.WebView2.Core
Imports Microsoft.Web.WebView2.WinForms
Imports System
Imports System.Drawing
Imports System.IO
Imports System.Windows.Forms

Namespace DevExpresss.DashboardWin.CustomItemExtension
	Public Class WebPageItemControlProvider
		Inherits CustomControlProviderBase

		Private imageStream As New MemoryStream()
		Private imageIsReady As Boolean = False
		Private webControl As WebView2
		Private dashboardItem As CustomDashboardItem(Of WebPageItemMetadata)
		Protected Overrides ReadOnly Property Control() As Control
			Get
				Return webControl
			End Get
		End Property

		Public Sub New(ByVal dashboardItem As CustomDashboardItem(Of WebPageItemMetadata))
			Me.dashboardItem = dashboardItem
			webControl = New WebView2()
			AddHandler webControl.CoreWebView2InitializationCompleted, AddressOf WebControl_CoreWebView2InitializationCompleted
		End Sub
		Private Sub PrepareImageStream()
			imageStream = New MemoryStream()
			imageIsReady = False
			webControl.CoreWebView2.CapturePreviewAsync(CoreWebView2CapturePreviewImageFormat.Png, imageStream).ContinueWith(Sub(t) imageIsReady = True)
		End Sub
		Private Sub WebControl_CoreWebView2InitializationCompleted(ByVal sender As Object, ByVal e As Microsoft.Web.WebView2.Core.CoreWebView2InitializationCompletedEventArgs)
			webControl.CoreWebView2.Settings.AreDefaultContextMenusEnabled = False
			AddHandler webControl.CoreWebView2.NavigationCompleted, Sub(s, ee) PrepareImageStream()
			AddHandler webControl.SizeChanged, Sub(s, ee) PrepareImageStream()
		End Sub
		Protected Overrides Sub UpdateControl(ByVal customItemData As CustomItemData)
			Try
				Dim uriString As String = String.Empty
				Dim attributeValue As Object = Nothing
				If dashboardItem.Metadata.Attribute IsNot Nothing Then
					attributeValue = customItemData.GetFlatData().GetDisplayText(dashboardItem.Metadata.Attribute.UniqueId, 0)
				End If
				If dashboardItem.Metadata.URI IsNot Nothing Then
					uriString = String.Format(dashboardItem.Metadata.URI, attributeValue)
				ElseIf attributeValue IsNot Nothing Then
					uriString = attributeValue.ToString()
				End If
				webControl.Source = New Uri(uriString)
			Catch
				webControl.Source = New Uri("about:blank")
			End Try
		End Sub
		Protected Overrides Function GetPrintableControl(ByVal customItemData As CustomItemData, ByVal exportInfo As CustomItemExportInfo) As XRControl
			If imageIsReady Then
				Dim xRPictureBox As New XRPictureBox()
				Dim bitmap As New Bitmap(imageStream)
				xRPictureBox.ImageSource = New ImageSource(bitmap)
				Return xRPictureBox
			End If
			Dim label As New XRLabel()
			label.Text = "Image did not load.Try to execute the Export command again."
			Return label
		End Function
	End Class
End Namespace

File: CS/CustomItemExtension/CustomItems/WebPageItem/WebPageItemControlProvider.cs
using DevExpress.DashboardCommon;
using DevExpress.DashboardWin;
using DevExpress.XtraPrinting.Drawing;
using DevExpress.XtraReports.UI;
using Microsoft.Web.WebView2.Core;
using Microsoft.Web.WebView2.WinForms;
using System;
using System.Drawing;
using System.IO;
using System.Windows.Forms;

namespace DevExpresss.DashboardWin.CustomItemExtension
{
    public class WebPageItemControlProvider : CustomControlProviderBase {
        MemoryStream imageStream = new MemoryStream();
        bool imageIsReady = false;
        WebView2 webControl;
        CustomDashboardItem<WebPageItemMetadata> dashboardItem;
        protected override Control Control { get { return webControl; } }

        public WebPageItemControlProvider(CustomDashboardItem<WebPageItemMetadata> dashboardItem) {
            this.dashboardItem = dashboardItem;
            webControl = new WebView2();
            webControl.CoreWebView2InitializationCompleted += WebControl_CoreWebView2InitializationCompleted;
        }
        void PrepareImageStream()
        {
            imageStream = new MemoryStream();
            imageIsReady = false;
            webControl.CoreWebView2.CapturePreviewAsync(CoreWebView2CapturePreviewImageFormat.Png, imageStream).ContinueWith((t) => imageIsReady = true);
        }
        private void WebControl_CoreWebView2InitializationCompleted(object sender, Microsoft.Web.WebView2.Core.CoreWebView2InitializationCompletedEventArgs e) {
            webControl.CoreWebView2.Settings.AreDefaultContextMenusEnabled = false;
            webControl.CoreWebView2.NavigationCompleted += (s, ee) => PrepareImageStream();
            webControl.SizeChanged += (s, ee) => PrepareImageStream();
        }
        protected override void UpdateControl(CustomItemData customItemData) {
            try {
                string uriString = string.Empty;
                object attributeValue = null;
                if(dashboardItem.Metadata.Attribute != null) {
                    attributeValue = customItemData.GetFlatData().GetDisplayText(dashboardItem.Metadata.Attribute.UniqueId, 0);
                }
                if(dashboardItem.Metadata.URI != null) {
                    uriString = string.Format(dashboardItem.Metadata.URI, attributeValue);
                }
                else if(attributeValue != null) {
                    uriString = attributeValue.ToString();
                }
                webControl.Source = new Uri(uriString);
            }
            catch {
                webControl.Source = new Uri("about:blank");
            };
        }
        protected override XRControl GetPrintableControl(CustomItemData customItemData, CustomItemExportInfo exportInfo) {
            if(imageIsReady) {
                XRPictureBox xRPictureBox = new XRPictureBox();
                Bitmap bitmap = new Bitmap(imageStream);
                xRPictureBox.ImageSource = new ImageSource(bitmap);
                return xRPictureBox;
            }
            XRLabel label = new XRLabel();
            label.Text = "Image did not load.Try to execute the Export command again.";
            return label;
        }
    }
}
